### ISSUE TITLE
Performance Regression since PenaltyBlog v1.5 fails to converge with dixon_coles_weights

### ISSUE THREAD

**Issue by ssoulless** [martineastwood/penaltyblog#16]
## Summary
Since v1.5.0 (commit 13b6b83 - gradients implementation), PenaltyBlog experiences massive convergence failures (78% of fits) when using `dixon_coles_weights()`. The same code works perfectly in v1.4.1.

## Version Comparison

| Version | Convergence Success | Failures | Total Time | Notes |
|---------|-------------------|----------|------------|-------|
| 1.4.1   | 113/113 (100%)    | 0        | 0.23s      | ✅ Gradients disabled |
| 1.5.0   | 25/113 (22%)      | 88 (78%) | 25.55s     | ❌ Gradients enabled (commit 13b6b83) |
| 1.6.1   | 25/113 (22%)      | 88 (78%) | 25.13s     | ❌ Issue persists |

**All tests WITHOUT weights work perfectly in all versions (100% convergence, ~0.13s).**

## Environment

- Python: 3.13.5
- OS: macOS (Darwin 24.6.0)
- Test data: 400 synthetic Poisson-distributed matches

### Minimal Script

```python
#!/usr/bin/env python3
"""
Performance regression test: PenaltyBlog 1.4.1 vs 1.5.0 vs 1.6.1
Demonstrates convergence failures with dixon_coles_weights since v1.5.0
"""
import penaltyblog as pb
import pandas as pd
import numpy as np
import time
from typing import List

print("=" * 70)
print(f"Version: {pb.__version__}")
print("=" * 70)

# Generate synthetic Poisson data (400 matches over 3 seasons)
np.random.seed(42)
dates = pd.date_range("2020-01-01", periods=400, freq="3D")
teams = [f"Team {chr(65+i)}" for i in range(4)]

df = pd.DataFrame({
    "date": dates,
    "team_home": np.random.choice(teams, 400),
    "team_away": np.random.choice(teams, 400),
    "goals_home": np.random.poisson(1.5, 400),
    "goals_away": np.random.poisson(1.2, 400),
})

# Rolling window validation with dixon_coles_weights
def rolling_window_validation(df: pd.DataFrame, xi: float = 0.5) -> tuple:
    fit_times = []
    failures = []

    last_season = df.sort_values("date")["date"].iloc[-1] - pd.DateOffset(years=1)
    season_df = df[df["date"] >= last_season].copy()
    unique_dates = sorted(season_df["date"].unique())

    for i, current in enumerate(unique_dates):
        train = df[(df["date"] < current) &
                   (df["date"] >= current - pd.DateOffset(years=1))].copy()
        test = season_df[season_df["date"] == current].copy()

        if len(train) < 100 or test.empty:
            continue

        # Calculate Dixon-Coles time weights
        weights = pb.models.dixon_coles_weights(train["date"], xi)

        # Fit Poisson model with weights
        model = pb.models.PoissonGoalsModel(
            train.goals_home,
            train.goals_away,
            train.team_home,
            train.team_away,
            weights=weights
        )

        start = time.time()
        try:
            model.fit()

            # Make predictions on test set
            for _, row in test.iterrows():
                try:
                    _ = model.predict(row.team_home, row.team_away)
                except Exception:
                    pass

            fit_times.append(time.time() - start)
        except ValueError as e:
            if "Iteration limit reached" in str(e):
                failures.append(i)
                print(f"⚠️  Convergence failed at fold {i}: {e}")
            else:
                raise

    return fit_times, failures

# Run test
fit_times, failures = rolling_window_validation(df, xi=0.5)

# Report results
print(f"\nTotal folds: {len(fit_times) + len(failures)}")
print(f"Successful: {len(fit_times)} ({len(fit_times)*100/(len(fit_times)+len(failures)):.0f}%)")
print(f"Failed: {len(failures)} ({len(failures)*100/(len(fit_times)+len(failures)):.0f}%)")
print(f"Total time: {sum(fit_times):.2f}s")
if fit_times:
    print(f"Average time per successful fit: {np.mean(fit_times):.3f}s")

if failures:
    print(f"\n⚠️  WARNING: {len(failures)} folds failed to converge with default settings")
    print("   This did not occur in PenaltyBlog 1.4.1")

print("\n" + "=" * 70)
print("\n📈 EXPECTED PERFORMANCE BASELINE:")
print("   PenaltyBlog 1.4.1: ~0.23s total (100% convergence)")
print("   PenaltyBlog 1.5.0+: ~25s total (22% convergence)")
```

### How to Run

```bash
# Test v1.4.1 (baseline)
pip install penaltyblog==1.4.1
python reproduction_script.py
# Expected: 113 successful, 0 failed, 0.23s

# Test v1.5.0 (regression introduced)
pip install penaltyblog==1.5.0
python reproduction_script.py
# Actual: 25 successful, 88 failed, 25.55s

# Test v1.6.1 (persists)
pip install penaltyblog==1.6.1
python reproduction_script.py
# Actual: 25 successful, 88 failed, 25.13s
```

## Expected Behavior

`dixon_coles_weights()` should work as reliably in v1.5.0+ as it does in v1.4.1:
- ✅ 100% convergence with `maxiter=1000` (default)
- ✅ Fast execution (~0.2-0.5s for 113 fits)
- ✅ Consistent results across versions

## Actual Behavior (v1.5.0+)

When using `weights = pb.models.dixon_coles_weights(train['date'], xi=0.5)`:
- ❌ 78% of fits fail with `"Iteration limit reached"`
- ❌ The 22% that succeed take ~155ms each (vs ~2ms in v1.4.1)
- ❌ Total time: 25s vs 0.23s (108x slower)

**Without weights (`xi=0`)**: Everything works perfectly in all versions.

## Root Cause Analysis

The regression was introduced in **v1.5.0** with commit **13b6b83** ("feat: added gradient to zip, poisson, dixon coles and neg binom models"). Well I am not 100% sure that's the problematic commit.

### Hypothesis

The interaction between:
1. Analytical gradients (new in v1.5.0)
2. Weighted data via `dixon_coles_weights()`
3. Default `maxiter=1000`

causes the optimizer to fail to converge in most cases. The gradients implementation may be correct for unweighted data but appears incompatible with weighted optimization.

### Workaround 1: Disable gradients (defeats v1.5.0 purpose)
```python
model.fit(use_gradient=False)
# Works but loses the gradient optimization benefit
```

### Workaround 2: Increase iterations (very slow)
```python
model.fit(minimizer_options={"maxiter": 5000})
# Helps some cases but 10-20x slower
```

### Workaround 3: Don't use weights (loses functionality)
```python
# Use xi=0 (no weights)
# Works perfectly but defeats time-weighting purpose
```

**None of these are acceptable for production use.**

## Questions

1. Were gradients tested with weighted data during v1.5.0 development?
2. Is there a known issue with the gradient implementation + weights?
3. Should `maxiter` be automatically increased when weights are provided?
4. Are the gradients in `_gradient()` correctly accounting for the weight vector?

---
